# Why: Expose the host-only core entry points and compile the public header as
# C11 without introducing implementation changes.
diff --git a/engine/src/flutter/shell/platform/embedder/BUILD.gn b/engine/src/flutter/shell/platform/embedder/BUILD.gn
index de5589c8db0..18588df292b 100644
--- a/engine/src/flutter/shell/platform/embedder/BUILD.gn
+++ b/engine/src/flutter/shell/platform/embedder/BUILD.gn
@@ -68,6 +68,7 @@ template("embedder_source_set") {
       "embedder_external_view_embedder.h",
       "embedder_include.c",
       "embedder_include2.c",
+      "flutter_core_host_include.c",
       "embedder_layers.cc",
       "embedder_layers.h",
       "embedder_platform_message_response.cc",
@@ -214,7 +215,12 @@ embedder_source_set("embedder_as_internal_library") {
 }
 
 source_set("embedder_headers") {
-  public = [ "embedder.h" ]
+  public = [
+    "embedder.h",
+    "flutter_core_host.h",
+  ]
+
+  public_deps = [ "//flutter/runtime_spi:runtime_spi_c_api" ]
 
   public_configs = [ "//flutter:config" ]
 }
diff --git a/engine/src/flutter/shell/platform/embedder/flutter_core_host.h b/engine/src/flutter/shell/platform/embedder/flutter_core_host.h
new file mode 100644
index 00000000000..369d2ce19d8
--- /dev/null
+++ b/engine/src/flutter/shell/platform/embedder/flutter_core_host.h
@@ -0,0 +1,249 @@
+// Copyright 2026 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef FLUTTER_SHELL_PLATFORM_EMBEDDER_FLUTTER_CORE_HOST_H_
+#define FLUTTER_SHELL_PLATFORM_EMBEDDER_FLUTTER_CORE_HOST_H_
+
+#include "flutter/runtime_spi/include/flutter_runtime_spi.h"
+#include "flutter/shell/platform/embedder/embedder.h"
+
+#if defined(__cplusplus)
+extern "C" {
+#endif
+
+typedef struct _FlutterCoreEngine* FlutterCoreEngine;
+
+typedef struct {
+  size_t struct_size;
+  // Igual a `FlutterProjectArgs.assets_path`.
+  const char* assets_path;
+  // Igual a `FlutterProjectArgs.icu_data_path`.
+  const char* icu_data_path;
+  // Igual a `FlutterProjectArgs.command_line_argc`.
+  int command_line_argc;
+  // Igual a `FlutterProjectArgs.command_line_argv`.
+  const char* const* command_line_argv;
+  // Igual a `FlutterProjectArgs.platform_message_callback`.
+  FlutterPlatformMessageCallback platform_message_callback;
+  // Igual a `FlutterProjectArgs.update_semantics_node_callback`.
+  FlutterUpdateSemanticsNodeCallback update_semantics_node_callback;
+  // Igual a `FlutterProjectArgs.update_semantics_custom_action_callback`.
+  FlutterUpdateSemanticsCustomActionCallback
+      update_semantics_custom_action_callback;
+  // Igual a `FlutterProjectArgs.persistent_cache_path`.
+  const char* persistent_cache_path;
+  // Igual a `FlutterProjectArgs.is_persistent_cache_read_only`.
+  bool is_persistent_cache_read_only;
+  // Igual a `FlutterProjectArgs.vsync_callback`.
+  VsyncCallback vsync_callback;
+  // Igual a `FlutterProjectArgs.custom_task_runners`.
+  const FlutterCustomTaskRunners* custom_task_runners;
+  // Igual a `FlutterProjectArgs.compositor`.
+  const FlutterCompositor* compositor;
+  // Igual a `FlutterProjectArgs.compute_platform_resolved_locale_callback`.
+  FlutterComputePlatformResolvedLocaleCallback
+      compute_platform_resolved_locale_callback;
+  // Igual a `FlutterProjectArgs.on_pre_engine_restart_callback`.
+  OnPreEngineRestartCallback on_pre_engine_restart_callback;
+  // Igual a `FlutterProjectArgs.update_semantics_callback`.
+  FlutterUpdateSemanticsCallback update_semantics_callback;
+  // Igual a `FlutterProjectArgs.update_semantics_callback2`.
+  FlutterUpdateSemanticsCallback2 update_semantics_callback2;
+  // Igual a `FlutterProjectArgs.channel_update_callback`.
+  FlutterChannelUpdateCallback channel_update_callback;
+  // Igual a `FlutterProjectArgs.view_focus_change_request_callback`.
+  FlutterViewFocusChangeRequestCallback view_focus_change_request_callback;
+  // Igual a `FlutterProjectArgs.enable_wide_gamut`.
+  bool enable_wide_gamut;
+} FlutterCoreHostArgs;
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineInitialize(
+    size_t version,
+    const FlutterRendererConfig* config,
+    const FlutterCoreHostArgs* args,
+    const FlutterSpiRuntimeTable* runtime,
+    void* user_data,
+    FlutterCoreEngine* engine_out);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineRunInitialized(FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineRun(
+    size_t version,
+    const FlutterRendererConfig* config,
+    const FlutterCoreHostArgs* args,
+    const FlutterSpiRuntimeTable* runtime,
+    void* user_data,
+    FlutterCoreEngine* engine_out);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineDeinitialize(FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineShutdown(FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineAddView(FlutterCoreEngine engine,
+                                             const FlutterAddViewInfo* info);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineRemoveView(
+    FlutterCoreEngine engine,
+    const FlutterRemoveViewInfo* info);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendViewFocusEvent(
+    FlutterCoreEngine engine,
+    const FlutterViewFocusEvent* event);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendWindowMetricsEvent(
+    FlutterCoreEngine engine,
+    const FlutterWindowMetricsEvent* event);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendPointerEvent(
+    FlutterCoreEngine engine,
+    const FlutterPointerEvent* events,
+    size_t events_count);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendKeyEvent(
+    FlutterCoreEngine engine,
+    const FlutterKeyEvent* event,
+    FlutterKeyEventCallback callback,
+    void* user_data);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendPlatformMessage(
+    FlutterCoreEngine engine,
+    const FlutterPlatformMessage* message);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCorePlatformMessageCreateResponseHandle(
+    FlutterCoreEngine engine,
+    FlutterDataCallback data_callback,
+    void* user_data,
+    FlutterPlatformMessageResponseHandle** response_out);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCorePlatformMessageReleaseResponseHandle(
+    FlutterCoreEngine engine,
+    FlutterPlatformMessageResponseHandle* response);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendPlatformMessageResponse(
+    FlutterCoreEngine engine,
+    const FlutterPlatformMessageResponseHandle* handle,
+    const uint8_t* data,
+    size_t data_length);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineRegisterExternalTexture(
+    FlutterCoreEngine engine,
+    int64_t texture_identifier);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineUnregisterExternalTexture(
+    FlutterCoreEngine engine,
+    int64_t texture_identifier);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineMarkExternalTextureFrameAvailable(
+    FlutterCoreEngine engine,
+    int64_t texture_identifier);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineUpdateSemanticsEnabled(
+    FlutterCoreEngine engine,
+    bool enabled);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineUpdateAccessibilityFeatures(
+    FlutterCoreEngine engine,
+    FlutterAccessibilityFeature features);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineDispatchSemanticsAction(
+    FlutterCoreEngine engine,
+    uint64_t node_id,
+    FlutterSemanticsAction action,
+    const uint8_t* data,
+    size_t data_length);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSendSemanticsAction(
+    FlutterCoreEngine engine,
+    const FlutterSendSemanticsActionInfo* info);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineOnVsync(FlutterCoreEngine engine,
+                                             intptr_t baton,
+                                             uint64_t frame_start_time_nanos,
+                                             uint64_t frame_target_time_nanos);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineReloadSystemFonts(FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+void FlutterCoreTraceEventDurationBegin(const char* name);
+
+FLUTTER_EXPORT
+void FlutterCoreTraceEventDurationEnd(const char* name);
+
+FLUTTER_EXPORT
+void FlutterCoreTraceEventInstant(const char* name);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEnginePostRenderThreadTask(
+    FlutterCoreEngine engine,
+    VoidCallback callback,
+    void* callback_data);
+
+FLUTTER_EXPORT
+uint64_t FlutterCoreGetCurrentTime();
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineRunTask(FlutterCoreEngine engine,
+                                             const FlutterTask* task);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineUpdateLocales(
+    FlutterCoreEngine engine,
+    const FlutterLocale** locales,
+    size_t locales_count);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineNotifyLowMemoryWarning(
+    FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEnginePostCallbackOnAllNativeThreads(
+    FlutterCoreEngine engine,
+    FlutterNativeThreadCallback callback,
+    void* user_data);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineNotifyDisplayUpdate(
+    FlutterCoreEngine engine,
+    FlutterEngineDisplaysUpdateType update_type,
+    const FlutterEngineDisplay* displays,
+    size_t display_count);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineScheduleFrame(FlutterCoreEngine engine);
+
+FLUTTER_EXPORT
+FlutterEngineResult FlutterCoreEngineSetNextFrameCallback(
+    FlutterCoreEngine engine,
+    VoidCallback callback,
+    void* user_data);
+
+#if defined(__cplusplus)
+}  // extern "C"
+#endif
+
+#endif  // FLUTTER_SHELL_PLATFORM_EMBEDDER_FLUTTER_CORE_HOST_H_
diff --git a/engine/src/flutter/shell/platform/embedder/flutter_core_host_include.c b/engine/src/flutter/shell/platform/embedder/flutter_core_host_include.c
new file mode 100644
index 00000000000..72debd9efef
--- /dev/null
+++ b/engine/src/flutter/shell/platform/embedder/flutter_core_host_include.c
@@ -0,0 +1,8 @@
+// Copyright 2026 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "flutter/shell/platform/embedder/flutter_core_host.h"
+
+// This file is only present to ensure that the public core host header can be
+// cleanly included in a C11 translation unit.
